pkg/util/podstore/podstore.go

```diff
@@ -0,0 +1,231 @@
+		s.timestampMap[pod.UID] = &podTimestamps{CreationTimestamp: timeNow}
+	default:
+		s.timestampMap[pod.UID] = &podTimestamps{CreationTimestamp: pod.CreationTimestamp.Time}
+	}
+	klog.V(4).InfoS("Processed Pod Create Event", "Pod", klog.KObj(pod))
+}
+
+func (s *PodStore) onPodDelete(obj interface{}) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	timeNow := s.clock.Now()
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		var err error
+		pod, err = s.checkDeletedPod(obj)
+		if err != nil {
+			klog.ErrorS(err, "Got error while processing Delete Event")
+			return
+		}
+	}
+	timestamp, ok := s.timestampMap[pod.UID]
+	if !ok {
+		klog.ErrorS(nil, "Cannot find podTimestamps in timestampMap", "UID", pod.UID)
+		return
+	}
+	timestamp.DeletionTimestamp = &timeNow
+	s.podsToDelete.AddAfter(pod, delayTime)
+	klog.V(4).InfoS("Processed Pod Delete Event", "Pod", klog.KObj(pod))
+}
+
+func (s *PodStore) checkDeletedPod(obj interface{}) (*corev1.Pod, error) {
+	deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		return nil, fmt.Errorf("received unexpected object: %v", obj)
+	}
+	pod, ok := deletedState.Obj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("DeletedFinalStateUnknown object is not of type Pod: %v", deletedState.Obj)
+	}
+	return pod, nil
+}
+
+func (s *PodStore) GetPodByIPAndTime(ip string, time time.Time) (*corev1.Pod, bool) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	pods, _ := s.pods.ByIndex(podIPIndex, ip)
+	if len(pods) == 0 {
+		return nil, false
+	} else if len(pods) == 1 {
+		pod := pods[0].(*corev1.Pod)
+		// In case the clocks may be skewed between different Nodes in the cluster, we directly return the Pod if there is only
+		// one Pod in the indexer. Otherwise, we check the timestamp for Pods in the indexer.
+		klog.V(4).InfoS("Matched Pod IP to Pod from indexer", "ip", ip, "Pod", klog.KObj(pod))
+		return pod, true
+	}
+	for _, pod := range pods {
+		pod := pod.(*corev1.Pod)
+		timestamp, ok := s.timestampMap[pod.UID]
+		if !ok {
+			continue
+		}
+		if timestamp.CreationTimestamp.Before(time) && (timestamp.DeletionTimestamp == nil || time.Before(*timestamp.DeletionTimestamp)) {
+			klog.V(4).InfoS("Matched Pod IP and time to Pod from indexer", "ip", ip, "time", time, "Pod", klog.KObj(pod))
+			return pod, true
+		}
+	}
+	return nil, false
+}
+
+func (s *PodStore) Run(stopCh <-chan struct{}) {
+	defer s.podsToDelete.ShutDown()
+	go wait.Until(s.worker, time.Second, stopCh)
+	<-stopCh
+}
+
+// worker runs a worker thread that just dequeues item from deleteQueue and
+// remove the item from prevPod.
+func (s *PodStore) worker() {
+	for s.processDeleteQueueItem() {
+	}
+}
+
+func (s *PodStore) processDeleteQueueItem() bool {
+	pod, quit := s.podsToDelete.Get()
+	if quit {
+		return false
+	}
+	err := s.pods.Delete(pod)
+	if err != nil {
+		klog.ErrorS(err, "Error when deleting Pod from deletion workqueue", "Pod", klog.KObj(pod.(*corev1.Pod)))
+		return false
+	}
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
```

Comment:
It updates `s.pods` and `s.timestampMap` atomically in other places but not here. I didn't think of a negative case for now but we should be careful about such inconsistency.

Reply:
The reason I put the Lock here is because `s.podsToDelete.Get()` will wait until we have Pod added in to the delayQueue, which causes deadlock here. But I should put the lock before `s.pods.Delete(pod)`.